File: dotnet/src/OpenBankingIO.Client/OpenBankingClient.cs
using System.Globalization;
using System.Net.Http.Json;
using System.Security.Cryptography;
using System.Text.Json;

namespace OpenBankingIO.Client;

/// <summary>
/// Server-to-server client for open-banking.io. Authenticates with an API key (<c>X-Api-Key</c>) and
/// decrypts the zero-knowledge data envelopes locally with the exported private key — the service only
/// ever returns ciphertext it cannot read.
/// </summary>
public sealed class OpenBankingClient : IDisposable
{
    private readonly HttpClient _http;
    private readonly ECDiffieHellman _privateKey;
    private readonly bool _ownsHttp;

    /// <param name="apiBaseUrl">e.g. <c>https://open-banking.io</c>.</param>
    /// <param name="apiKey">The API key from your credentials bundle.</param>
    /// <param name="privateKeyPkcs8Base64">The base64 PKCS#8 encryption private key from your bundle.</param>
    /// <param name="httpClient">Optional injected client (e.g. for testing); otherwise one is created.</param>
    public OpenBankingClient(string apiBaseUrl, string apiKey, string privateKeyPkcs8Base64, HttpClient? httpClient = null)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(apiBaseUrl);
        ArgumentException.ThrowIfNullOrWhiteSpace(apiKey);
        ArgumentException.ThrowIfNullOrWhiteSpace(privateKeyPkcs8Base64);

        _privateKey = ECDiffieHellman.Create();
        _privateKey.ImportPkcs8PrivateKey(Convert.FromBase64String(privateKeyPkcs8Base64), out _);

        _ownsHttp = httpClient is null;
        _http = httpClient ?? new HttpClient();
        _http.BaseAddress = new Uri(apiBaseUrl.TrimEnd('/') + "/");
        _http.DefaultRequestHeaders.Remove("X-Api-Key");
        _http.DefaultRequestHeaders.Add("X-Api-Key", apiKey);
    }

    /// <summary>Builds a client from a credentials-bundle JSON string or a path to a bundle file.</summary>
    public static OpenBankingClient FromCredentials(string jsonOrPath, HttpClient? httpClient = null)
    {
        var json = File.Exists(jsonOrPath) ? File.ReadAllText(jsonOrPath) : jsonOrPath;
        var bundle = JsonSerializer.Deserialize<CredentialsBundle>(json, Json.Options)
            ?? throw new ArgumentException("Could not parse the credentials bundle", nameof(jsonOrPath));
        return FromBundle(bundle, httpClient);
    }

    public static OpenBankingClient FromBundle(CredentialsBundle bundle, HttpClient? httpClient = null)
    {
        if (string.IsNullOrWhiteSpace(bundle.ApiKey))
            throw new ArgumentException("The credentials bundle has no apiKey", nameof(bundle));
        return new OpenBankingClient(bundle.ApiBaseUrl, bundle.ApiKey!, bundle.EncryptionKey.PrivateKey, httpClient);
    }

    /// <summary>Lists the user's accounts with all sensitive fields decrypted.</summary>
    public async Task<IReadOnlyList<Account>> GetAccountsAsync(CancellationToken ct = default)
    {
        var wires = await GetAccountWiresAsync(ct).ConfigureAwait(false);
        return wires.Select(MapAccount).ToList();
    }

    /// <summary>Returns a page of an account's statement, newest first, with decrypted fields.</summary>
    public async Task<TransactionPage> GetTransactionsAsync(
        string accountId, DateOnly? from = null, DateOnly? to = null, int? limit = null, int? offset = null,
        CancellationToken ct = default)
    {
        var query = new List<string>();
        if (from is { } f) query.Add($"from={f:yyyy-MM-dd}");
        if (to is { } t) query.Add($"to={t:yyyy-MM-dd}");
        if (limit is { } l) query.Add($"limit={l}");
        if (offset is { } o) query.Add($"offset={o}");
        var qs = query.Count > 0 ? "?" + string.Join("&", query) : "";

        var page = await _http.GetFromJsonAsync<TransactionPageWire>(
            $"api/accounts/{accountId}/transactions{qs}", Json.Options, ct).ConfigureAwait(false)
            ?? new TransactionPageWire();

        return new TransactionPage { Items = page.Items.Select(MapTransaction).ToList(), Total = page.Total };
    }

    /// <summary>Lists the user's bank connections.</summary>
    public async Task<IReadOnlyList<Connection>> GetConnectionsAsync(CancellationToken ct = default)
    {
        var wires = await _http.GetFromJsonAsync<List<ConnectionWire>>("api/connections", Json.Options, ct)
            .ConfigureAwait(false) ?? [];
        return wires.Select(c => new Connection
        {
            SessionId = c.SessionId,
            AspspName = c.AspspName,
            AspspCountry = c.AspspCountry,
            ValidUntil = c.ValidUntil,
            Status = c.Status,
            AccountCount = c.AccountCount,
            LastSyncedAt = c.LastSyncedAt,
            PsuType = c.PsuType,
        }).ToList();
    }

    /// <summary>
    /// Triggers an online sync of one account: decrypts that account's Enable Banking uid and posts it,
    /// so the service can fetch fresh data without ever holding the uid in plaintext.
    /// </summary>
    public async Task<SyncResult> SyncAsync(string accountId, CancellationToken ct = default)
    {
        var wires = await GetAccountWiresAsync(ct).ConfigureAwait(false);
        var account = wires.FirstOrDefault(a => a.Id == accountId)
            ?? throw new InvalidOperationException($"Account {accountId} not found");
        var uid = DecryptUid(account)
            ?? throw new InvalidOperationException("Account has no active session (reconnect required) — cannot sync");

        var response = await _http.PostAsJsonAsync($"api/accounts/{accountId}/sync", new { uid }, Json.Options, ct)
            .ConfigureAwait(false);
        response.EnsureSuccessStatusCode();
        var result = await response.Content.ReadFromJsonAsync<SyncResultWire>(Json.Options, ct).ConfigureAwait(false)
            ?? new SyncResultWire(0, 0);
        return new SyncResult(result.NewTransactions, result.TotalFetched);
    }

    /// <summary>Triggers an online sync of every account that has an active session.</summary>
    public async Task<SyncAllResult> SyncAllAsync(CancellationToken ct = default)
    {
        var wires = await GetAccountWiresAsync(ct).ConfigureAwait(false);
        var items = wires
            .Select(a => (a.Id, Uid: DecryptUid(a)))
            .Where(x => x.Uid is not null)
            .Select(x => new { accountId = x.Id, uid = x.Uid })
            .ToList();

        var response = await _http.PostAsJsonAsync("api/sync", new { items }, Json.Options, ct).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();
        var result = await response.Content.ReadFromJsonAsync<SyncAllResultWire>(Json.Options, ct).ConfigureAwait(false)
            ?? new SyncAllResultWire(0, 0);
        return new SyncAllResult(result.Accounts, result.NewTransactions);
    }

    private async Task<List<AccountWire>> GetAccountWiresAsync(CancellationToken ct)
        => await _http.GetFromJsonAsync<List<AccountWire>>("api/accounts", Json.Options, ct).ConfigureAwait(false) ?? [];

    private string? DecryptUid(AccountWire a)
        => Envelope.DecryptTo<UidEnc>(_privateKey, a.UidEnc)?.Uid;

    private Account MapAccount(AccountWire a)
    {
        var acc = Envelope.DecryptTo<AccountEnc>(_privateKey, a.Enc);
        var name = Envelope.DecryptTo<DisplayNameEnc>(_privateKey, a.DisplayNameEnc);

        return new Account
        {
            Id = a.Id,
            AspspName = a.AspspName,
            AspspCountry = a.AspspCountry,
            Currency = a.Currency,
            AccountType = a.AccountType,
            Bic = a.Bic,
            NeedsReconnect = a.NeedsReconnect,
            Iban = acc?.Iban,
            Bban = acc?.Bban,
            OwnerName = acc?.OwnerName,
            AccountName = acc?.AccountName,
            Product = acc?.Product,
            DisplayName = name?.DisplayName,
            Balances = a.Balances.Select(b =>
            {
                var dec = Envelope.DecryptTo<BalanceEnc>(_privateKey, b.Enc);
                return new Balance
                {
                    Type = b.Type,
                    Currency = b.Currency,
                    ReferenceDate = b.ReferenceDate,
                    Name = dec?.Name,
                    Amount = ParseDecimal(dec?.Amount),
                };
            }).ToList(),
        };
    }

    private Transaction MapTransaction(TransactionWire t)
    {
        var d = Envelope.DecryptTo<TransactionEnc>(_privateKey, t.Enc);
        return new Transaction
        {
            Id = t.Id,
            Currency = t.Currency,
            CreditDebitIndicator = t.CreditDebitIndicator,
            Status = t.Status,
            BookingDate = t.BookingDate,
            ValueDate = t.ValueDate,
            TransactionDate = t.TransactionDate,
            BankTransactionCode = t.BankTransactionCode,
            Amount = ParseDecimal(d?.Amount),
            CreditorName = d?.CreditorName,
            CreditorIban = d?.CreditorIban,
            CreditorBban = d?.CreditorBban,
            CreditorAgentBic = d?.CreditorAgentBic,
            DebtorName = d?.DebtorName,
            DebtorIban = d?.DebtorIban,
            DebtorBban = d?.DebtorBban,
            RemittanceInformation = d?.RemittanceInformation,
            Note = d?.Note,
            ReferenceNumber = d?.ReferenceNumber,
            ExchangeRate = d?.ExchangeRate,
            MerchantCategoryCode = d?.MerchantCategoryCode,
            BalanceAfterCurrency = d?.BalanceAfterCurrency,
            BalanceAfterTransaction = ParseDecimalNullable(d?.BalanceAfter),
        };
    }

    private static decimal ParseDecimal(string? s)
        => string.IsNullOrEmpty(s) ? 0m : decimal.Parse(s, CultureInfo.InvariantCulture);

    private static decimal? ParseDecimalNullable(string? s)
        => string.IsNullOrEmpty(s) ? null : decimal.Parse(s, CultureInfo.InvariantCulture);

    public void Dispose()
    {
        _privateKey.Dispose();
        if (_ownsHttp) _http.Dispose();
    }
}

File: java/src/main/java/io/openbanking/OpenBankingClient.java
package io.openbanking;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.math.BigDecimal;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.interfaces.ECPrivateKey;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Server-to-server client for open-banking.io. Authenticates with an API key ({@code X-Api-Key})
 * and decrypts the zero-knowledge data envelopes locally with the exported private key — the
 * service only ever returns ciphertext it cannot read.
 */
public final class OpenBankingClient {

  /** Connect timeout for the default HTTP client. */
  private static final Duration CONNECT_TIMEOUT = Duration.ofSeconds(10);

  /** Per-request timeout applied to every request this client sends. */
  private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(30);

  /** User-Agent sent on every request; {@code <version>} comes from the JAR manifest. */
  private static final String USER_AGENT = "open-banking-io/java/" + version();

  private final String baseUrl;
  private final String apiKey;
  private final ECPrivateKey privateKey;
  private final HttpClient http;
  private final ObjectMapper mapper =
      new ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

  /**
   * @param apiBaseUrl e.g. {@code https://open-banking.io}
   * @param apiKey the API key from your credentials bundle
   * @param privateKeyPkcs8Base64 the base64 PKCS#8 encryption private key from your bundle
   * @param httpClient an injected client (e.g. for testing), or {@code null} for a default one
   */
  public OpenBankingClient(
      String apiBaseUrl, String apiKey, String privateKeyPkcs8Base64, HttpClient httpClient) {
    if (isBlank(apiBaseUrl)) {
      throw new OpenBankingException("apiBaseUrl is required");
    }
    if (isBlank(apiKey)) {
      throw new OpenBankingException("apiKey is required");
    }
    if (isBlank(privateKeyPkcs8Base64)) {
      throw new OpenBankingException("privateKeyPkcs8 is required");
    }
    this.baseUrl = apiBaseUrl.replaceAll("/+$", "");
    this.apiKey = apiKey;
    this.privateKey = Envelope.loadPrivateKey(privateKeyPkcs8Base64);
    this.http =
        httpClient != null
            ? httpClient
            : HttpClient.newBuilder().connectTimeout(CONNECT_TIMEOUT).build();
  }

  /**
   * The implementation version baked into the JAR manifest by Maven, or {@code "dev"} when running
   * from a build tree without a manifest (e.g. tests).
   */
  private static String version() {
    String v = OpenBankingClient.class.getPackage().getImplementationVersion();
    return (v == null || v.isBlank()) ? "dev" : v;
  }

  /** Builds a client from a credentials-bundle JSON string or a path to a bundle file. */
  public static OpenBankingClient fromCredentials(String jsonOrPath) {
    return fromCredentials(jsonOrPath, null);
  }

  public static OpenBankingClient fromCredentials(String jsonOrPath, HttpClient httpClient) {
    String raw = jsonOrPath;
    if (!jsonOrPath.stripLeading().startsWith("{")) {
      try {
        raw = Files.readString(Path.of(jsonOrPath));
      } catch (IOException e) {
        throw new OpenBankingException("could not read credentials file", e);
      }
    }
    CredentialsBundle bundle;
    try {
      bundle =
          new ObjectMapper()
              .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
              .readValue(raw, CredentialsBundle.class);
    } catch (IOException e) {
      throw new OpenBankingException("could not parse the credentials bundle", e);
    }
    return fromBundle(bundle, httpClient);
  }

  public static OpenBankingClient fromBundle(CredentialsBundle bundle, HttpClient httpClient) {
    if (bundle == null || isBlank(bundle.apiKey())) {
      throw new OpenBankingException("the credentials bundle has no apiKey");
    }
    if (bundle.encryptionKey() == null || isBlank(bundle.encryptionKey().privateKey())) {
      throw new OpenBankingException("the credentials bundle has no encryption private key");
    }
    return new OpenBankingClient(
        bundle.apiBaseUrl(), bundle.apiKey(), bundle.encryptionKey().privateKey(), httpClient);
  }

  /** Lists the user's accounts with all sensitive fields decrypted. */
  public List<Account> getAccounts() {
    List<Account> accounts = new ArrayList<>();
    for (AccountWire w : getAccountWires()) {
      accounts.add(mapAccount(w));
    }
    return accounts;
  }

  /** Returns a page of an account's statement, newest first, with decrypted fields. */
  public TransactionPage getTransactions(String accountId, TransactionQuery query) {
    if (query == null) {
      query = TransactionQuery.none();
    }
    StringBuilder qs = new StringBuilder();
    appendParam(qs, "from", query.from());
    appendParam(qs, "to", query.to());
    appendParam(qs, "limit", query.limit() == null ? null : query.limit().toString());
    appendParam(qs, "offset", query.offset() == null ? null : query.offset().toString());

    String path = "/api/accounts/" + encodePathSegment(accountId) + "/transactions" + qs;
    TransactionPageWire page = getJson(path, TransactionPageWire.class);

    List<Transaction> items = new ArrayList<>();
    List<TransactionWire> wireItems = page.items() == null ? List.of() : page.items();
    for (TransactionWire t : wireItems) {
      items.add(mapTransaction(t));
    }
    return new TransactionPage(items, page.total());
  }

  /** Lists the user's bank connections. */
  public List<Connection> getConnections() {
    List<ConnectionWire> wires = getJsonList("/api/connections", ConnectionWire.class);
    List<Connection> conns = new ArrayList<>();
    for (ConnectionWire c : wires) {
      conns.add(
          new Connection(
              c.sessionId(),
              c.aspspName(),
              c.aspspCountry(),
              c.validUntil(),
              c.status(),
              c.accountCount(),
              c.lastSyncedAt(),
              c.psuType()));
    }
    return conns;
  }

  /**
   * Triggers an online sync of one account: decrypts that account's Enable Banking uid and posts
   * it, so the service can fetch fresh data without ever holding the uid in plaintext.
   */
  public SyncResult sync(String accountId) {
    AccountWire account = null;
    for (AccountWire a : getAccountWires()) {
      if (a.id().equals(accountId)) {
        account = a;
        break;
      }
    }
    if (account == null) {
      throw new OpenBankingException("account " + accountId + " not found");
    }
    String uid = decryptUid(account);
    if (uid == null) {
      throw new OpenBankingException(
          "account has no active session (reconnect required) — cannot sync");
    }

    SyncResultWire result =
        postJson(
            "/api/accounts/" + encodePathSegment(accountId) + "/sync",
            Map.of("uid", uid),
            SyncResultWire.class);
    return new SyncResult(result.newTransactions(), result.totalFetched());
  }

  /** Triggers an online sync of every account that has an active session. */
  public SyncAllResult syncAll() {
    List<Map<String, String>> items = new ArrayList<>();
    for (AccountWire a : getAccountWires()) {
      String uid = decryptUid(a);
      if (uid != null) {
        Map<String, String> item = new LinkedHashMap<>();
        item.put("accountId", a.id());
        item.put("uid", uid);
        items.add(item);
      }
    }

    SyncAllResultWire result =
        postJson("/api/sync", Map.of("items", items), SyncAllResultWire.class);
    return new SyncAllResult(result.accounts(), result.newTransactions());
  }

  // ---- internals ----

  private List<AccountWire> getAccountWires() {
    return getJsonList("/api/accounts", AccountWire.class);
  }

  private String decryptUid(AccountWire a) {
    UidEnc dec = Envelope.decryptTo(privateKey, a.uidEnc(), mapper, UidEnc.class);
    return dec == null ? null : dec.uid();
  }

  private Account mapAccount(AccountWire a) {
    AccountEnc acc = Envelope.decryptTo(privateKey, a.enc(), mapper, AccountEnc.class);
    DisplayNameEnc name =
        Envelope.decryptTo(privateKey, a.displayNameEnc(), mapper, DisplayNameEnc.class);

    List<Balance> balances = new ArrayList<>();
    List<BalanceWire> wireBalances = a.balances() == null ? List.of() : a.balances();
    for (BalanceWire b : wireBalances) {
      BalanceEnc dec = Envelope.decryptTo(privateKey, b.enc(), mapper, BalanceEnc.class);
      balances.add(
          new Balance(
              b.type(),
              dec == null ? null : dec.name(),
              parseDecimal(dec == null ? null : dec.amount()),
              b.currency(),
              b.referenceDate()));
    }

    return new Account(
        a.id(),
        a.aspspName(),
        a.aspspCountry(),
        a.currency(),
        a.accountType(),
        a.bic(),
        a.needsReconnect(),
        acc == null ? null : acc.iban(),
        acc == null ? null : acc.bban(),
        acc == null ? null : acc.ownerName(),
        acc == null ? null : acc.accountName(),
        acc == null ? null : acc.product(),
        name == null ? null : name.displayName(),
        balances);
  }

  private Transaction mapTransaction(TransactionWire t) {
    TransactionEnc d = Envelope.decryptTo(privateKey, t.enc(), mapper, TransactionEnc.class);
    return new Transaction(
        t.id(),
        t.currency(),
        t.creditDebitIndicator(),
        t.status(),
        t.bookingDate(),
        t.valueDate(),
        t.transactionDate(),
        t.bankTransactionCode(),
        parseDecimal(d == null ? null : d.amount()),
        d == null ? null : d.creditorName(),
        d == null ? null : d.creditorIban(),
        d == null ? null : d.creditorBban(),
        d == null ? null : d.creditorAgentBic(),
        d == null ? null : d.debtorName(),
        d == null ? null : d.debtorIban(),
        d == null ? null : d.debtorBban(),
        d == null ? null : d.debtorAgentBic(),
        d == null ? null : d.remittanceInformation(),
        d == null ? null : d.note(),
        d == null ? null : d.referenceNumber(),
        d == null ? null : d.exchangeRate(),
        d == null ? null : d.merchantCategoryCode(),
        parseDecimalNullable(d == null ? null : d.balanceAfter()),
        d == null ? null : d.balanceAfterCurrency());
  }

  private <T> T getJson(String path, Class<T> type) {
    return readJson(sendGet(path), path, mapper.getTypeFactory().constructType(type));
  }

  private <T> List<T> getJsonList(String path, Class<T> elementType) {
    JavaType listType = mapper.getTypeFactory().constructCollectionType(List.class, elementType);
    return readJson(sendGet(path), path, listType);
  }

  private byte[] sendGet(String path) {
    HttpRequest req =
        HttpRequest.newBuilder(URI.create(baseUrl + path))
            .header("X-Api-Key", apiKey)
            .header("User-Agent", USER_AGENT)
            .timeout(REQUEST_TIMEOUT)
            .GET()
            .build();
    return send(req, "GET", path);
  }

  private <T> T postJson(String path, Object body, Class<T> type) {
    String payload;
    try {
      payload = mapper.writeValueAsString(body);
    } catch (IOException e) {
      throw new OpenBankingException("could not serialize request body", e);
    }
    HttpRequest req =
        HttpRequest.newBuilder(URI.create(baseUrl + path))
            .header("X-Api-Key", apiKey)
            .header("Content-Type", "application/json")
            .header("User-Agent", USER_AGENT)
            .timeout(REQUEST_TIMEOUT)
            .POST(HttpRequest.BodyPublishers.ofString(payload, StandardCharsets.UTF_8))
            .build();
    byte[] data = send(req, "POST", path);
    return readJson(data, path, mapper.getTypeFactory().constructType(type));
  }

  private byte[] send(HttpRequest req, String method, String path) {
    HttpResponse<byte[]> resp;
    try {
      resp = http.send(req, HttpResponse.BodyHandlers.ofByteArray());
    } catch (IOException e) {
      throw new OpenBankingException(method + " " + path + " failed", e);
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new OpenBankingException(method + " " + path + " was interrupted", e);
    }
    if (resp.statusCode() < 200 || resp.statusCode() >= 300) {
      throw new OpenBankingException(method + " " + path + " failed: HTTP " + resp.statusCode());
    }
    return resp.body();
  }

  private <T> T readJson(byte[] data, String path, JavaType type) {
    try {
      return mapper.readValue(data, type);
    } catch (IOException e) {
      throw new OpenBankingException(path + " returned an unparseable body", e);
    }
  }

  private static void appendParam(StringBuilder qs, String key, String value) {
    if (value == null) {
      return;
    }
    qs.append(qs.length() == 0 ? '?' : '&')
        .append(key)
        .append('=')
        .append(URLEncoder.encode(value, StandardCharsets.UTF_8));
  }

  private static String encodePathSegment(String s) {
    return URLEncoder.encode(s, StandardCharsets.UTF_8).replace("+", "%20");
  }

  private static BigDecimal parseDecimal(String s) {
    return (s == null || s.isEmpty()) ? BigDecimal.ZERO : new BigDecimal(s);
  }

  private static BigDecimal parseDecimalNullable(String s) {
    return (s == null || s.isEmpty()) ? null : new BigDecimal(s);
  }

  private static boolean isBlank(String s) {
    return s == null || s.isBlank();
  }
}
